Python/remote_debugging.c

```diff
@@ -0,0 +1,939 @@
+        return -1;
+    }
+
+    if (!_Py_Debug_Free_Threaded && debug_offsets->free_threaded) {
+        PyErr_SetString(
+            PyExc_RuntimeError,
+            "Cannot send commands to a free-threaded Python process"
+            " from a process running a non-free-threaded version");
+        return -1;
+    }
+
+    return 0;
+}
+
+static int
+read_offsets(
+    proc_handle_t *handle,
+    uintptr_t *runtime_start_address,
+    _Py_DebugOffsets* debug_offsets
+) {
+    *runtime_start_address = get_py_runtime(handle);
+    if (!*runtime_start_address) {
+        if (!PyErr_Occurred()) {
+            PyErr_SetString(
+                PyExc_RuntimeError, "Failed to get PyRuntime address");
+        }
+        return -1;
+    }
+    size_t size = sizeof(struct _Py_DebugOffsets);
+    if (0 != read_memory(handle, *runtime_start_address, size, debug_offsets)) {
+        return -1;
+    }
+    if (ensure_debug_offset_compatibility(debug_offsets)) {
+        return -1;
+    }
+    return 0;
+}
+
+static int
+send_exec_to_proc_handle(proc_handle_t *handle, int tid, const char *debugger_script_path)
+{
+    uintptr_t runtime_start_address;
+    struct _Py_DebugOffsets debug_offsets;
+
+    if (read_offsets(handle, &runtime_start_address, &debug_offsets)) {
+        return -1;
+    }
+
+    uintptr_t interpreter_state_list_head = (uintptr_t)debug_offsets.runtime_state.interpreters_head;
+
+    uintptr_t interpreter_state_addr;
+    if (0 != read_memory(
+            handle,
+            runtime_start_address + interpreter_state_list_head,
+            sizeof(void*),
+            &interpreter_state_addr))
+    {
+        return -1;
+    }
+
+    if (interpreter_state_addr == 0) {
+        PyErr_SetString(PyExc_RuntimeError, "Can't find a running interpreter in the remote process");
+        return -1;
+    }
+
+    int is_remote_debugging_enabled = 0;
+    if (0 != read_memory(
+            handle,
+            interpreter_state_addr + debug_offsets.debugger_support.remote_debugging_enabled,
+            sizeof(int),
+            &is_remote_debugging_enabled))
+    {
+        return -1;
+    }
+
+    if (is_remote_debugging_enabled != 1) {
+        PyErr_SetString(
+            PyExc_RuntimeError,
+            "Remote debugging is not enabled in the remote process");
+        return -1;
+    }
+
+    uintptr_t thread_state_addr;
+    unsigned long this_tid = 0;
+
+    if (tid != 0) {
+        if (0 != read_memory(
+                handle,
+                interpreter_state_addr + debug_offsets.interpreter_state.threads_head,
+                sizeof(void*),
+                &thread_state_addr))
+        {
+            return -1;
+        }
+        while (thread_state_addr != 0) {
+            if (0 != read_memory(
+                    handle,
+                    thread_state_addr + debug_offsets.thread_state.native_thread_id,
+                    sizeof(pid_t),
+                    &this_tid))
```

Comment:
Size mismatch? Should `this_tid` be defined as `pid_t` instead of `unsigned long` above (and change the cast below)

Reply:
Ooh, size mismatch indeed. No, `this_tid` should be an `unsigned long`, but the bug is having `sizeof(pid_t)` here instead of `sizeof(this_tid)`. Nice catch.